File: go/004/main.go
/*
 * ProjectEuler 004
 * Go / main.go
 * copyright (c) 2015 Susisu
 *
 * Go 1.4.1
 * `go build -o bin/main main.go`
 */


package main

import (
    "fmt"
)

func main() {
    max := 0
    for i := 100; i < 1000; i++ {
        for j := 100; j < 1000; j++ {
            p := i * j;
            if isPalinfrome(p) && p > max {
                max = p
            }
        }
    }
    fmt.Printf("%d\n", max)
}

func isPalinfrome(n int) bool {
    s := fmt.Sprintf("%d", n)
    l := len(s)
    h := l / 2
    for i := 0; i < h; i++ {
        if s[i] != s[l - 1 - i] {
            return false;
        }
    }
    return true;
}

File: ruby/004/main.rb
=begin
    ProjectEuler 004
    Ruby / main.rb
    copyright (c) 2015 Susisu

    Ruby 2.0.0p481
    `ruby main.rb`
=end

def main
    max = 0
    for i in 100 .. 999
        for j in i .. 999
            p = i * j
            if (palindrome? p) && p > max then
                max = p
            end
        end
    end
    puts max
end

def palindrome?(n)
    s = n.to_s
    return s == s.reverse
end

main()
